### PR TITLE
Add DEBUG messages into cuDNN conv codes.

### LINKED ISSUE #6157: After updating Theano weird problems when using GPU
Recently I've updated Theano to the master branch. However, this lead to the following error:
`
Exception: ('The following error happened while compiling the node', GpuDnnConvGradI{algo='none', inplace=True}(GpuContiguous.0, GpuContiguous.0, GpuAllocEmpty{dtype='float32', context_name=None}.0, GpuDnnConvDesc{border_mode='half', subsample=(1, 1), dilation=(1, 1), conv_mode='conv', precision='float32'}.0, Constant{1.0}, Constant{0.0}), '\n', "Compilation failed (return status=1): /home/abotev/.theano/compiledir_Linux-2.6-el6.x86_64-x86_64-with-centos-6.5-Final-x86_64-3.6.0-64/tmpa8wrg7og/mod.cpp:442: error: a brace-enclosed initializer is not allowed here before '{' token. /home/abotev/.theano/compiledir_Linux-2.6-el6.x86_64-x86_64-with-centos-6.5-Final-x86_64-3.6.0-64/tmpa8wrg7og/mod.cpp:442: error: ISO C++ forbids initialization of member 'prev_kern_dims'. /home/abotev/.theano/compiledir_Linux-2.6-el6.x86_64-x86_64-with-centos-6.5-Final-x86_64-3.6.0-64/tmpa8wrg7og/mod.cpp:442: error: making 'prev_kern_dims' static. /home/abotev/.theano/compiledir_Linux-2.6-el6.x86_64-x86_64-with-centos-6.5-Final-x86_64-3.6.0-64/tmpa8wrg7og/mod.cpp:442: error: invalid in-class initialization of static data member of non-integral type 'size_t [5]'. /home/abotev/.theano/compiledir_Linux-2.6-el6.x86_64-x86_64-with-centos-6.5-Final-x86_64-3.6.0-64/tmpa8wrg7og/mod.cpp:443: error: a brace-enclosed initializer is not allowed here before '{' token. /home/abotev/.theano/compiledir_Linux-2.6-el6.x86_64-x86_64-with-centos-6.5-Final-x86_64-3.6.0-64/tmpa8wrg7og/mod.cpp:443: error: ISO C++ forbids initialization of member 'prev_top_dims'. /home/abotev/.theano/compiledir_Linux-2.6-el6.x86_64-x86_64-with-centos-6.5-Final-x86_64-3.6.0-64/tmpa8wrg7og/mod.cpp:443: error: making 'prev_top_dims' static. /home/abotev/.theano/compiledir_Linux-2.6-el6.x86_64-x86_64-with-centos-6.5-Final-x86_64-3.6.0-64/tmpa8wrg7og/mod.cpp:443: error: invalid in-class initialization of static data member of non-integral type 'size_t [5]'. /home/abotev/.theano/compiledir_Linux-2.6-el6.x86_64-x86_64-with-centos-6.5-Final-x86_64-3.6.0-64/tmpa8wrg7og/mod.cpp: In member function 'int<unnamed>::__struct_compiled_op_mc2bb2e24bb41d8af45c9d8be40d95ddd::conv_gi_node_mc2bb2e24bb41d8af45c9d8be40d95ddd_0(PyGpuArrayObject*, PyGpuArrayObject*, PyGpuArrayObject*, cudnnConvolutionStruct*, double, double, PyGpuArrayObject**, _Params_c626b80a28594f6fe623577946e29157_c29b5bb55c42161a5f82e1c2559749fa*)':. /home/abotev/.theano/compiledir_Linux-2.6-el6.x86_64-x86_64-with-centos-6.5-Final-x86_64-3.6.0-64/tmpa8wrg7og/mod.cpp:558: error: 'prev_kern_dims' was not declared in this scope. /home/abotev/.theano/compiledir_Linux-2.6-el6.x86_64-x86_64-with-centos-6.5-Final-x86_64-3.6.0-64/tmpa8wrg7og/mod.cpp:560: error: 'prev_top_dims' was not declared in this scope. /home/abotev/.theano/compiledir_Linux-2.6-el6.x86_64-x86_64-with-centos-6.5-Final-x86_64-3.6.0-64/tmpa8wrg7og/mod.cpp:625: error: 'prev_kern_dims' was not declared in this scope. /home/abotev/.theano/compiledir_Linux-2.6-el6.x86_64-x86_64-with-centos-6.5-Final-x86_64-3.6.0-64/tmpa8wrg7og/mod.cpp:626: error: 'prev_top_dims' was not declared in this scope. /home/abotev/.theano/compiledir_Linux-2.6-el6.x86_64-x86_64-with-centos-6.5-Final-x86_64-3.6.0-64/tmpa8wrg7og/mod.cpp: In member function 'int<unnamed>::__struct_compiled_op_mc2bb2e24bb41d8af45c9d8be40d95ddd::init(PyObject*, PyObject*, PyObject*, PyObject*, PyObject*, PyObject*, PyObject*, PyObject*, PyObject*)':. /home/abotev/.theano/compiledir_Linux-2.6-el6.x86_64-x86_64-with-centos-6.5-Final-x86_64-3.6.0-64/tmpa8wrg7og/mod.cpp:883: error: 'prev_kern_dims' was not declared in this scope. /home/abotev/.theano/compiledir_Linux-2.6-el6.x86_64-x86_64-with-centos-6.5-Final-x86_64-3.6.0-64/tmpa8wrg7og/mod.cpp:884: error: 'prev_top_dims' was not declared in this scope. ", "[GpuDnnConvGradI{algo='none', inplace=True}(<GpuArrayType<None>(float32, (True, False, False, False))>, <GpuArrayType<None>(float32, (False, True, False, False))>, <GpuArrayType<None>(float32, 4D)>, <CDataType{cudnnConvolutionDescriptor_t}>, Constant{1.0}, Constant{0.0})]")
`
I've also tried updating `libgpuarray` to the latest but that did not resolve it either. The code however runs fine on the CPU and worked fine before on the GPU. 

It seems that the error is coming from the C code of the kernel `prev_kern_dims` where there is a mistake of a brace initializer. However, I have no idea where exactly to look for that.

### PR DESCRIPTION
**Pull request by notoraptor**
Also:
- Fix docstring for c_init_code_struct.
- Update support_code_struct for dnn_gi ( try to fix #6157 based on comment https://github.com/Theano/Theano/issues/6157#issuecomment-316088272 ).

Should I add any kind of tests ? @lamblin @abergeron .

### CODE REVIEW

File: theano/gpuarray/dnn_gi.c [hunk 1]
```diff
@@ -13,8 +13,8 @@ if (PARAMS->choose_algo) {
 
 int reuse_algo;
 cudnnConvolutionBwdDataAlgo_t prev_algo;
-size_t prev_kern_dims[5] = {0};
-size_t prev_top_dims[5] = {0};
+size_t prev_kern_dims[5];
+size_t prev_top_dims[5];
```
Comment (abergeron) [inline theano/gpuarray/dnn_gi.c]: We still need to initialize those to all zeros.  But this can be done in init_code struct().

Comment (notoraptor) [inline theano/gpuarray/dnn_gi.c]: It is indeed done ! Here: https://github.com/notoraptor/Theano/blob/ee22927e1e74001df7b115339ab1d5e482b6fb23/theano/gpuarray/dnn_gi.c#L6

Comment (lamblin) [inline theano/gpuarray/dnn_gi.c]: I would say we should probably init them all the time, just for safety.

Comment (notoraptor) [inline theano/gpuarray/dnn_gi.c]: Ok. In that case, I will have to update the three files (dnn_fwd, dnn_gw, dnn_gi), as they have currently the same policy about these arrays.

File: theano/gpuarray/dnn_fwd.c [hunk 2]
```diff
@@ -138,6 +141,16 @@ APPLY_SPECIFIC(conv_fwd)(PyGpuArrayObject *input, PyGpuArrayObject *kerns,
           return 1;
         }
         algo = choice.algo;
+
+        #ifdef DEBUG
+        if (count == 0) {
+            PyErr_SetString(PyExc_RuntimeError, "No best-timed conv fwd algorithm found");
+            return 1;
+        } else {
+            fprintf(stderr, "(%d best-timed conv fwd algorithms)\n", count);
```
Comment (lamblin) [inline theano/gpuarray/dnn_fwd.c]: Printing the content of the `cudnnConvolutionFwdAlgoPerf_t` that is returned in `choice` as well would be nice.
In fact, there should probably be an error if `algo.status != CUDNN_STATUS_SUCCESS`, like if `count == 0`.

File: theano/gpuarray/dnn_gi.c [hunk 3]
```diff
@@ -178,6 +181,16 @@ APPLY_SPECIFIC(conv_gi)(PyGpuArrayObject *kerns, PyGpuArrayObject *output,
...
 
       algo = choice.algo;
+
+      #ifdef DEBUG
+      if (count == 0) {
+          PyErr_SetString(PyExc_RuntimeError, "No best-timed conv gradinput algorithm found");
+          return 1;
+      } else {
+          fprintf(stderr, "(%d best-timed conv gradinput algorithms)\n", count);
+      }
```
Comment (lamblin) [inline theano/gpuarray/dnn_gi.c]: Same thing for here.

File: theano/gpuarray/dnn_gw.c [hunk 4]
```diff
@@ -180,6 +183,16 @@ APPLY_SPECIFIC(conv_gw)(PyGpuArrayObject *input, PyGpuArrayObject *output,
...
 
         algo = choice.algo;
+
+        #ifdef DEBUG
+        if (count == 0) {
+            PyErr_SetString(PyExc_RuntimeError, "No best-timed conv gradweight algorithm found");
+            return 1;
+        } else {
+            fprintf(stderr, "(%d best-timed conv gradweight algorithms)\n", count);
+        }
```
Comment (lamblin) [inline theano/gpuarray/dnn_gw.c]: and here

File: theano/gpuarray/dnn_gi.c (hunk 1 above)
Comment (lamblin) [inline theano/gpuarray/dnn_gi.c]: Yes, if you could just do the memset outside of the `if` for all three files, that would be great.

**Comment by notoraptor** [comment]
Updated @abergeron @lamblin .

**Comment by notoraptor** [comment]
Jenkins aborted here: http://earlgrey.iro.umontreal.ca:8080/job/Theano_PR/457/console

jenkins test this

**Comment by notoraptor** [comment]
Jenkins succeeded.

**Pull request merged**